chore: update CredentialProvider

poppoerika opened the pull request:
- Added `StringMomentoTokenProvider`
- Added static functions, `fromEnvironmentVariable` and `fromString` 
- Added some test to test out the above

Closes #100

malandis commented:
Looks good. Can you document the `StringMomentoProvider` class and `fromEnvironmentVariable` method as well?

pgautier404 commented:
This looks good but there is a lot of duplicated code between the two providers. In other SDKs we had the `EnvMomentoTokenProvider` inherit from `StringMomentoTokenProvider` since the only extra thing it does is read a stering from an environment variable. I think doing the same here would simplify and make this more maintainable. The node SDK is a good model for this pattern.

poppoerika commented:
> This looks good but there is a lot of duplicated code between the two providers. In other SDKs we had the `EnvMomentoTokenProvider` inherit from `StringMomentoTokenProvider` since the only extra thing it does is read a stering from an environment variable. I think doing the same here would simplify and make this more maintainable. The node SDK is a good model for this pattern.

@pgautier404 
I wanted to discuss this with you, so sounds good!

cprice404 commented on src/Auth/ICredentialProvider.php:
```diff
@@ -1,17 +0,0 @@
-
-namespace Momento\Auth;
-
-interface ICredentialProvider
```
why delete the interface?

pgautier404 replied:
Yeah, we kind of went back and forth on this but I think @cprice404 is right. Keeping the interface around gives us extra flexibility, and that's definitely what we should be returning from the factory methods, @poppoerika.

cprice404 reviewed (commented):
i feel like it's probably still a good idea to keep the interface around and use it in the return types from the factory functions, but will defer to Pete.

pgautier404 commented on src/Auth/CredentialProvider.php:
```diff
@@ -0,0 +1,30 @@
+ * Provides information that the CacheClient needs in order to establish a connection to and authenticate with
+ * the Momento service.
+ */
+abstract class CredentialProvider
```
```suggestion
abstract class CredentialProvider implements ICredentialProvider
```

We'll of course need to add the `ICredentialProvider` file and class back in.

pgautier404 commented on src/Auth/EnvMomentoTokenProvider.php:
```diff
@@ -27,42 +23,17 @@ public function __construct(
+     * @return EnvMomentoTokenProvider
+     * @throws InvalidArgumentError
+     */
+    public static function fromEnvironmentVariable(string $envVariableName): EnvMomentoTokenProvider
```
This function should be in the `CredentialProvider` class and should return `ICredentialProvider` type.

pgautier404 commented on src/Auth/StringMomentoTokenProvider.php:
```diff
@@ -0,0 +1,74 @@
+     * @return StringMomentoTokenProvider
+     * @throws InvalidArgumentError
+     */
+    public static function fromString(string $authToken): StringMomentoTokenProvider
```
Please move this function to `CredentialProvider` and mark it as returning an `ICredentialProvider`.

pgautier404 commented on src/Cache/CacheClient.php:
```diff
@@ -48,11 +48,11 @@ class CacheClient implements LoggerAwareInterface
      */
     public function __construct(
-        IConfiguration $configuration, ICredentialProvider $authProvider, int $defaultTtlSeconds
+        IConfiguration $configuration, CredentialProvider $authProvider, int $defaultTtlSeconds
```
```suggestion
        IConfiguration $configuration, ICredentialProvider $authProvider, int $defaultTtlSeconds
```

pgautier404 replied:
Sorry about the runaround, but we'll want these types to be `ICredentialProvider` pretty much everywhere in the clients and grpc managers.

poppoerika commented:
@pgautier404 @cprice404 
Thank you for the feedback! 
After talking to Pete, it makes sense to keep `ICredentialProvider` and implement it in `CredentialProvider` to add flexibility to our interface.
I added back `ICredentialProvider` and updated `CredentialProvier` accordingly. 
Also updated the examples so that we use `CacheClient` and `CredentialProvider::fromEnvironmentVariable`.
Ah, also updated `CredentialProvider` return type to `ICredentialProvier` type.

pgautier404 commented on src/Cache/Psr16CacheClient.php:
```diff
@@ -26,15 +26,15 @@ class Psr16CacheClient implements CacheInterface
     /**
      * @param IConfiguration $configuration
-     * @param ICredentialProvider $authProvider
+     * @param CredentialProvider $authProvider
```
```suggestion
     * @param ICredentialProvider $authProvider
```

pgautier404 commented on src/Cache/Psr16CacheClient.php:
```diff
@@ -26,15 +26,15 @@ class Psr16CacheClient implements CacheInterface
-        ?int                $defaultTtlSeconds,
-        ?bool               $throwExceptions = null
+        IConfiguration     $configuration,
+        CredentialProvider $authProvider,
```
```suggestion
        ICredentialProvider $authProvider,
```

pgautier404 commented on src/Auth/CredentialProvider.php:
```diff
@@ -0,0 +1,41 @@
+     * @return StringMomentoTokenProvider
+     * @throws InvalidArgumentError
+     */
+    public static function fromString(string $authToken): StringMomentoTokenProvider
```
```suggestion
    public static function fromString(string $authToken): ICredentialProvider
```

pgautier404 commented on src/Auth/CredentialProvider.php:
```diff
@@ -0,0 +1,41 @@
+     * @return EnvMomentoTokenProvider
+     * @throws InvalidArgumentError
+     */
+    public static function fromEnvironmentVariable(string $envVariableName): EnvMomentoTokenProvider
```
```suggestion
    public static function fromEnvironmentVariable(string $envVariableName): ICredentialProvider
```

pgautier404 reviewed (approved):
LGTM!
